File: java-impl/src/main/java/net/terria/registry/terraform/ProviderController.java
package net.terria.registry.terraform;

import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/v1/providers")
public class ProviderController {

    @Autowired
    private RegistryService registryService;
    
    private static final String DEFAULT_REGISTRY = "registry.terraform.io";

    /**
     * Get provider versions endpoint.
     * Terraform Registry Protocol: GET /v1/providers/{namespace}/{type}/versions
     */
    @GetMapping("/{namespace}/{name}/versions")
    public ResponseEntity<?> getProviderVersions(@PathVariable String namespace,
                                                 @PathVariable String name) {
        return registryService.getProvider(DEFAULT_REGISTRY, namespace, name)
            .map(index -> {
                List<Map<String, Object>> versions = index.versions.entrySet().stream()
                    .map(entry -> Map.of(
                        "version", entry.getKey().toString(),
                        "platforms", entry.getValue().stream()
                            .map(p -> Map.of("os", p.os, "arch", p.arch))
                            .collect(Collectors.toList())
                    ))
                    .collect(Collectors.toList());

                return ResponseEntity.ok(Map.of("versions", versions));
            })
            .orElse(ResponseEntity.notFound().build());
    }

    /**
     * Download metadata endpoint (redirect target can be mapped externally).
     * Terraform Registry Protocol: GET /v1/providers/{namespace}/{type}/{version}/download/{os}/{arch}
     */
    @GetMapping("/{namespace}/{name}/{version}/download/{os}/{arch}")
    public ResponseEntity<?> downloadProvider(@PathVariable String namespace,
                                              @PathVariable String name,
                                              @PathVariable String version,
                                              @PathVariable String os,
                                              @PathVariable String arch) {
        String url = String.format("/storage/providers/%s/%s/%s/%s/%s/terraform-provider-%s_%s_%s_%s.zip",
            namespace, name, version, os, arch, name, version, os, arch);
        return ResponseEntity.ok(Map.of("download_url", url));
    }

}

File: dotnet-impl/main/net/terria/registry/terraform/ProviderController.cs
using Microsoft.AspNetCore.Mvc;

namespace net.terria.registry.terraform
{
    [ApiController]
    [Route("v1/providers")]
    public class ProviderController : ControllerBase
    {
        private readonly RegistryService _registryService;
        private const string DefaultRegistry = "registry.terraform.io";

        public ProviderController(RegistryService registryService)
        {
            _registryService = registryService;
        }

        /// <summary>
        /// Get provider versions endpoint.
        /// Terraform Registry Protocol: GET /v1/providers/{namespace}/{name}/versions
        /// </summary>
        [HttpGet("{namespace}/{name}/versions")]
        public IActionResult GetProviderVersions(string @namespace, string name)
        {
            var index = _registryService.GetProvider(DefaultRegistry, @namespace, name);
            if (index == null) return NotFound();

            var versions = index.versions
                .OrderByDescending(v => v.Key)
                .Select(v => new
                {
                    version = v.Key.ToString(),
                    platforms = v.Value.Select(p => new { os = p.os, arch = p.arch })
                });

            return Ok(new { versions });
        }

        /// <summary>
        /// Download metadata endpoint (redirect target can be mapped externally).
        /// Terraform Registry Protocol: GET /v1/providers/{namespace}/{name}/{version}/download/{os}/{arch}
        /// </summary>
        [HttpGet("{namespace}/{name}/{version}/download/{os}/{arch}")]
        public IActionResult DownloadProvider(string @namespace, string name, string version, string os, string arch)
        {
            var url = $"/storage/providers/{@namespace}/{name}/{version}/{os}/{arch}/terraform-provider-{name}_{version}_{os}_{arch}.zip";
            return Ok(new { download_url = url });
        }
    }
}